### PR TITLE
Adapt distri pattern for Leap 16.0

### PR DESCRIPTION
**Pull request by lemon-suse**
- Related ticket: https://progress.opensuse.org/issues/167087
- Needles: N/A
- Verification run: 
  https://openqa.opensuse.org/tests/4514146#step/agama_auto/1 agama-default-auto on x86_64
  https://openqa.opensuse.org/tests/4514164#   agama-default on aarch64

### CODE REVIEW

File: lib/DistributionProvider.pm [hunk 1]
```diff
@@ -39,9 +39,9 @@ sub provide {
     return Distribution::Sle::15sp2->new() if is_sle('>15');
     return Distribution::Sle::15sp0->new() if is_sle('=15');
     return Distribution::Sle::12->new() if is_sle('12+');
+    return Distribution::Opensuse::AgamaDevel->new() if is_leap('=16.0');
```
Comment (jknphy) [inline lib/DistributionProvider.pm:42]: 16.0+

File: lib/DistributionProvider.pm [hunk 2]
```diff
@@ -39,9 +39,9 @@ sub provide {
     return Distribution::Sle::15sp2->new() if is_sle('>15');
     return Distribution::Sle::15sp0->new() if is_sle('=15');
     return Distribution::Sle::12->new() if is_sle('12+');
+    return Distribution::Opensuse::AgamaDevel->new() if is_leap('16.0+');
     return Distribution::Opensuse::Leap::15->new() if is_leap('15.0+');
     return Distribution::Opensuse::Leap::42->new() if is_leap('42.0+');
-    return Distribution::Opensuse::AgamaDevel->new() if get_var('VERSION', '') =~ /agama/;
```
Comment (jknphy) [inline lib/DistributionProvider.pm:44]: you cannot remove this by the way.

File: lib/DistributionProvider.pm [hunk 3]
```diff
@@ -39,9 +39,9 @@ sub provide {
     return Distribution::Sle::15sp2->new() if is_sle('>15');
     return Distribution::Sle::15sp0->new() if is_sle('=15');
     return Distribution::Sle::12->new() if is_sle('12+');
+    return Distribution::Opensuse::AgamaDevel->new() if is_leap('16.0+');
```
Comment (jknphy) [inline lib/DistributionProvider.pm:42]: actually you need to implement Sle::16.0:: package, I noticed that reading your pr title and commit that is incorrect, this is not related with booting but with "Adapt distri pattern for Lea 16.0"

File: lib/DistributionProvider.pm (hunk 2 above)
Comment (lemon-suse) [reply lib/DistributionProvider.pm:44]: Recovered it.

File: lib/DistributionProvider.pm (hunk 3 above)
Comment (lemon-suse) [reply lib/DistributionProvider.pm:42]: Updated the pr title and commit.
So you mean add another line as Sle::16.0::package to support Leap 16.0 on OSD?

File: lib/DistributionProvider.pm [hunk 4]
```diff
@@ -39,6 +40,7 @@ sub provide {
     return Distribution::Sle::15sp2->new() if is_sle('>15');
     return Distribution::Sle::15sp0->new() if is_sle('=15');
     return Distribution::Sle::12->new() if is_sle('12+');
+    return Distribution::Sle::16_current->new() if is_leap('16.0+');
```
Comment (jknphy) [inline lib/DistributionProvider.pm:43]: s/16_current/16Latest/

File: lib/DistributionProvider.pm (hunk 1 above)
Comment (lemon-suse) [reply lib/DistributionProvider.pm]: Updated.

File: lib/Distribution/Sle/16Latest.pm [hunk 5]
```diff
@@ -0,0 +1,53 @@
+# SUSE's openQA tests
+#
+# Copyright 2019-2022 SUSE LLC
```
Comment (jknphy) [inline lib/Distribution/Sle/16Latest.pm:3]: please, update.

File: lib/Distribution/Sle/16Latest.pm [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+# SPDX-License-Identifier: FSFAP
+
+# Summary: The class represents current (i.e. latest) Sle15 distribution and
+# provides access to its features.
+# Follows the "Factory first" rule. So that the feature first appears in
+# Tumbleweed distribution, and only if it behaves different in Sle15 then it
+# should be overriden here.
```
Comment (jknphy) [inline lib/Distribution/Sle/16Latest.pm:10]: This is copy/paste, doesn't apply.

File: lib/Distribution/Sle/16Latest.pm [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+sub get_reboot_page {
+    if (is_s390x()) {
+        return Yam::Agama::Pom::RebootTextmodePage->new();
+    }
+    else {
+        return Yam::Agama::Pom::RebootPage->new();
+    }
+
```
Comment (jknphy) [inline lib/Distribution/Sle/16Latest.pm:49]: blank line.

File: lib/Distribution/Sle/16Latest.pm (hunk 5 above)
Comment (lemon-suse) [reply lib/Distribution/Sle/16Latest.pm]: updated

File: lib/Distribution/Sle/16Latest.pm (hunk 7 above)
Comment (lemon-suse) [reply lib/Distribution/Sle/16Latest.pm]: removed.

File: lib/Distribution/Sle/16Latest.pm [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+1;
+
```
[the remark this answers is not in the document]
Comment (lemon-suse) [reply lib/Distribution/Sle/16Latest.pm]: removed

File: lib/Distribution/Sle/16Latest.pm [hunk 9]
```diff
@@ -0,0 +1,48 @@
+# SUSE's openQA tests
+#
+# Copyright 2024 SUSE LLC
+# SPDX-License-Identifier: FSFAP
+
+# Summary: Agama Devel distribution (not really a real distribution!)
+# for integration tests using the current state of code at GitHub for Agama.
```
Comment (jknphy) [inline lib/Distribution/Sle/16Latest.pm:7]: It is not this really, it is a real distri here, but doesn't inherit from other (for now). You can try some sort description.

File: lib/Distribution/Sle/16Latest.pm (hunk 6 above)
Comment (lemon-suse) [reply lib/Distribution/Sle/16Latest.pm]: copied from agama devel, not sure fit or not.

File: lib/Distribution/Sle/16Latest.pm (hunk 9 above)
Comment (jknphy) [reply lib/Distribution/Sle/16Latest.pm:7]: +# Summary: The class represents current (i.e. latest) Sle16 distribution and
+# provides access to its features.

File: lib/Distribution/Sle/16Latest.pm [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+# Maintainer: QE YaST and Migration (QE Yam) <qe-yam at suse de>
+
+package Distribution::Sle::16Latest;
+use strict;
+use warnings FATAL => 'all';
+use parent 'susedistribution';
```
Comment (jknphy) [inline lib/Distribution/Sle/16Latest.pm:14]: ci detects duplicate code, what we can do for now is to inherit from devel here so you don't need to rewrite all.

File: lib/Distribution/Sle/16Latest.pm (hunk 9 above)
Comment (lemon-suse) [reply lib/Distribution/Sle/16Latest.pm]: updated.

File: lib/Distribution/Sle/16Latest.pm [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+use warnings FATAL => 'all';
+use parent Distribution::Opensuse::AgamaDevel;
+
+use Yam::Agama::Pom::GrubMenuPage;
+use Yam::Agama::Pom::GrubEntryEditionPage;
+use Yam::Agama::Pom::AgamaUpAndRunningPage;
+use Yam::Agama::Pom::RebootPage;
+use Yam::Agama::Pom::RebootTextmodePage;
+
+use Utils::Architectures;
```
Comment (jknphy) [inline lib/Distribution/Sle/16Latest.pm:22]: you don't need any of this either....

File: lib/Distribution/Sle/16Latest.pm (hunk 10 above)
Comment (lemon-suse) [reply lib/Distribution/Sle/16Latest.pm]: Updated.

File: lib/Distribution/Sle/16Latest.pm (hunk 11 above)
Comment (lemon-suse) [reply lib/Distribution/Sle/16Latest.pm]: removed, a VR test is running https://openqa.opensuse.org/tests/4514637#live

File: lib/Distribution/Sle/16Latest.pm [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+
+# Summary: The class represents current (i.e. latest) SLE 16 distribution and
+# provides access to its features.
+
+# Maintainer: QE YaST and Migration (QE Yam) <qe-yam at suse de>
+
+package Distribution::Sle::16Latest;
+use parent Distribution::Opensuse::AgamaDevel;
+use strict;
+use warnings FATAL => 'all';
```
Comment (jknphy) [inline lib/Distribution/Sle/16Latest.pm:14]: and these two lines above maybe they are not needed either, right?

Comment (lemon-suse) [reply lib/Distribution/Sle/16Latest.pm]: yes, updated.

Comment (jknphy) [reply lib/Distribution/Sle/16Latest.pm]: I think strict , we cannot remove it, I see errors.

**Comment by jknphy** [comment]
https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/20276 merged.

File: lib/Distribution/Sle/16Latest.pm (hunk 12 above)
Comment (lemon-suse) [reply lib/Distribution/Sle/16Latest.pm:14]: put them back,  the VR: https://openqa.opensuse.org/tests/4516024#details

**Pull request merged**